File: camerakit/src/main/java/public/types/com/camerakit/CameraSize.java
package com.camerakit;


import android.hardware.Camera;

public class CameraSize implements Comparable<CameraSize> {

    private final int mWidth;
    private final int mHeight;

    public CameraSize(int width, int height) {
        mWidth = width;
        mHeight = height;
    }

    public int getWidth() {
        return mWidth;
    }

    public int getHeight() {
        return mHeight;
    }

    public CameraSize inverse() {
        return new CameraSize(mHeight, mWidth);
    }

    @Override
    public boolean equals(Object o) {
        if (o == null) {
            return false;
        }
        if (this == o) {
            return true;
        }
        if (o instanceof CameraSize) {
            CameraSize size = (CameraSize) o;
            return mWidth == size.mWidth && mHeight == size.mHeight;
        }
        return false;
    }

    @Override
    public String toString() {
        return mWidth + "x" + mHeight;
    }

    @Override
    public int hashCode() {
        return mHeight ^ ((mWidth << (Integer.SIZE / 2)) | (mWidth >>> (Integer.SIZE / 2)));
    }

    @Override
    public int compareTo(CameraSize another) {
        return mWidth * mHeight - another.mWidth * another.mHeight;
    }

    public static CameraSize get(Camera.Size size) {
        return new CameraSize(size.width, size.height);
    }

}

File: camerakit/src/main/java/camerakit/android/util/CameraSize.kt
package camerakit.android.util

import android.support.annotation.RequiresApi

class CameraSize(val width: Int, val height: Int) : Comparable<CameraSize> {

    companion object {

        operator fun invoke(size: android.hardware.Camera.Size): CameraSize {
            return CameraSize(size.width, size.height)
        }

        @RequiresApi(21)
        operator fun invoke(size: android.util.Size): CameraSize {
            return CameraSize(size.width, size.height)
        }

    }

    fun area(): Int {
        return width * height
    }

    fun aspectRatio(): Float {
        return width.toFloat() / height.toFloat()
    }

    fun asLandscape(): CameraSize {
        return when (width >= height) {
            true -> this
            false -> CameraSize(height, width)
        }
    }

    fun asPortrait(): CameraSize {
        return when (height >= width) {
            true -> this
            false -> CameraSize(height, width)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (other is CameraSize) {
            return width == other.width
                    && height == other.height
        }

        return false
    }

    override fun toString(): String {
        return "${width}x${height}"
    }

    override fun hashCode(): Int {
        return 31 * (31 + width.hashCode()) + height.hashCode()
    }

    override fun compareTo(other: CameraSize): Int {
        if (area() > other.area()) {
            return 1
        } else if (area() < other.area()) {
            return -1
        } else {
            return 0
        }
    }

}
